### PR TITLE
Card PR

### PR DESCRIPTION
**Pull request by saksham695**

### CODE REVIEW

File: examples/src/components/card.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+import ButtonComponent from './button';
+import CheckboxComponent from './Checkbox';
+import IconComponent from './icon';
+import ImageComponent from './image';
+import {theme} from '../index';
+import {applyTheme} from '../utility/utils';
+
+export default class CardComponent extends Component {
+  selectComponent(component, index) {
+    switch (component.type) {
```
Comment (mahima-vs) [inline examples/src/components/card.js:16]: No need to add another switch in card.js, as we already have a component base switch in ```index.js```

Comment (saksham695) [reply examples/src/components/card.js]: done

File: examples/src/components/card.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    );
+  }
+}
+
+CardComponent.propTypes = {
+  title: PropTypes.string,
+  containerStyle: PropTypes.shape({
+    color: PropTypes.string,
+    fontSize: PropTypes.number,
+    marginTop: PropTypes.number,
```
Comment (mahima-vs) [inline examples/src/components/card.js:37]: marginTop: PropTypes.number,
marginTop can be in string and in number as well (px or %)

File: examples/src/components/header.js [hunk 3]
```diff
@@ -25,8 +27,16 @@ export default class HeaderComponent extends React.Component {
       });
     }
   }
-
+  static contextType = NavigationContext;
+  componentDidMount() {
+    if (this.props.onPress) {
```
Comment (mahima-vs) [inline examples/src/components/header.js:32]: if (this.props.onPress) {
      this.props.onPress.navigation
Can we combine these 2 conditions  like- 
``` if (this.props.onPress &&  this.props.onPress.navigation) {
.....
}
```

File: examples/src/components/header.js [hunk 4]
```diff
@@ -41,13 +51,27 @@ export default class HeaderComponent extends React.Component {
             {this.props[component].childrens.map((item, i) => {
               if (item.type == 'icon') {
                 const styleOfIcon = item.properties;
+                if (item.properties.onPress) {
```
Comment (mahima-vs) [inline examples/src/components/header.js:54]: Similarly here.

File: examples/src/components/modal.js [hunk 5]
```diff
@@ -168,7 +150,7 @@ ModalComponent.propTypes = {
       backgroundColor: PropTypes.string,
       margin: PropTypes.number,
       marginBottom: PropTypes.number,
-      marginTop: PropTypes.number,
+      marginTop: PropTypes.string,
```
Comment (mahima-vs) [inline examples/src/components/modal.js:153]: marginTop can be number/string

File: examples/src/components/text.js [hunk 6]
```diff
@@ -12,18 +12,35 @@ import {Text} from 'react-native';
...
 
 export default class TextComponent extends React.Component {
   componentWillMount() {
     // Applying theme on text
     this.props = theme ? applyTheme(this.props, theme) : this.props;
   }
-
+  componentDidMount() {
+    if (this.props.onPress) {
+      this.props.onPress.navigation
```
Comment (mahima-vs) [inline examples/src/components/text.js:24]: Pls combine these 2 conditions.

File: examples/src/components/text.js [hunk 7]
```diff
@@ -12,18 +12,35 @@ import {Text} from 'react-native';
...
     const props = this.props;
     return (
       <>
-        <Text style={props.style}>{props.title}</Text>
+        <Text
+          style={props.style}
+          onPress={
+            this.props.onPress
+              ? () => navigation.navigate(this.props.onPress.screenName)
+              : () => console.log('Attach a Method')
```
Comment (mahima-vs) [inline examples/src/components/text.js:40]: Do not log "attach a method" here, add this in defaultProps and here, simply do whatever came from json. 
```
 this.props.onPress
              ? () => navigation.navigate(this.props.onPress.screenName)
              : () => this.props.onPress
```

File: examples/src/components/card.js (hunk 2 above)
Comment (saksham695) [reply examples/src/components/card.js:37]: done @mahima-vs

File: examples/src/components/header.js (hunk 3 above)
Comment (saksham695) [reply examples/src/components/header.js:32]: done

File: examples/src/components/card.js [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+import {Card, Button, Icon} from 'react-native-elements';
+import {View, ScrollView} from 'react-native';
+
+import {applyTheme} from '../utility/utils';
+import {theme} from '../index';
+
+export default class CardComponent extends Component {
+  constructor(props) {
+    super(props);
+    props = theme ? applyTheme(props, theme) : props;
```
Comment (mahima-vs) [inline examples/src/components/card.js:12]: props = theme ? applyTheme(props, theme) : props;
For more cleaner code, you can write 
```
if(theme){
applyTheme(props, theme)
}
```

File: examples/src/components/card.js [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    borderWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
+    padding: PropTypes.number,
+  }),
+};
+
+CardComponent.defaultProps = {
+  containerStyle: {
+    backgroundColor: 'white',
+
+    height: 300,
```
Comment (mahima-vs) [inline examples/src/components/card.js:55]: For defaultProps, pass some childrens as well, so tht if we won't pass any children in card, card still looks impressive.

File: examples/src/components/icon.js [hunk 10]
```diff
@@ -7,8 +7,12 @@ import {applyTheme} from '../utility/utils';
 import {theme} from '../index';
 
 export default class IconComponent extends Component {
-  render() {
+  constructor(props) {
+    super(props);
+
     this.props = theme ? applyTheme(this.props, theme) : this.props;
```
Comment (mahima-vs) [inline examples/src/components/icon.js:13]: Same here.

File: src/components/text.js [hunk 11]
```diff
@@ -6,47 +6,55 @@
...
-import { theme } from "../../index";
+import { theme } from '../../index';
 
 export default class TextComponent extends React.Component {
+
   componentWillMount() {
+
     // Applying theme on text
-    this.props = theme ? applyTheme(this.props, theme) : this.props;
+    this.props = theme ? applyTheme(this.props, theme) : this.props
```
Comment (mahima-vs) [inline src/components/text.js:21]: if(theme){
applyTheme(props, theme)
}

File: src/components/icon.js [hunk 12]
```diff
@@ -1,20 +1,14 @@
...
 export default class IconComponent extends Component {
-  constructor(props) {
-    super(props);
-    if (theme) {
-      // Applying theme on components.
-      props = applyTheme(props, theme);
-    }
-  }
   render() {
+    this.props = theme ? applyTheme(this.props, theme) : this.props;
```
Comment (mahima-vs) [inline src/components/icon.js:11]: Move applyTheme in constructor with 
if(theme){
applyTheme(props, theme)
}

File: src/components/card.js [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+import React, { Component } from "react";
+import { Card, Button, Icon } from "react-native-elements";
+import { View, ScrollView } from "react-native";
+
+import { applyTheme } from "../utility/utils";
+import { theme } from "../../index";
+
+export default class CardComponent extends Component {
+  constructor() {
+    this.props = theme ? applyTheme(this.props, theme) : this.props;
```
Comment (mahima-vs) [inline src/components/card.js:11]: if(theme){
applyTheme(props, theme)
}

File: examples/App.js [hunk 14]
```diff
@@ -8,12 +8,12 @@
 
 // import UiBuilder from 'react-native-ui-builder';
 import React from 'react';
-import UiBuilder from './src/navigation/navigation';
+import UiBuilder from './src/Navigation/navigation';
```
Comment (mahima-vs) [inline examples/App.js:11]: We are using small "n" not capital for folder naming convention.

File: examples/src/index.js [hunk 15]
```diff
@@ -7,21 +7,22 @@
...
 
 import React from 'react';
-import {View, KeyboardAvoidingView, ScrollView} from 'react-native';
 import {get} from 'lodash';
+import {View, KeyboardAvoidingView, ScrollView} from 'react-native';
 
 import {applyTheme} from './utility/utils';
 import ButtonComponent from './components/button';
 import CardComponent from './components/card';
+import CheckboxComponent from './components/card';
```
Comment (mahima-vs) [inline examples/src/index.js:16]: importing CheckboxComponent from card.js ?

File: examples/src/index.js [hunk 16]
```diff
@@ -105,6 +106,21 @@ export default class Components extends React.Component {
...
       case COMPONENTS.CHECKBOX:
         return <CheckboxComponent {...component.properties} key={index} />;
+
+      case COMPONENTS.CARD:
+        // map to store children components of card
+        const childComponents = component.childrens.map((component, index) => {
+          return this.selectComponent(component, index);
+        });
+
+        console.log(component);
```
Comment (mahima-vs) [inline examples/src/index.js:116]: remove irrelevant logging

File: examples/src/components/Checkbox.js [hunk 17]
```diff
@@ -1,64 +1,43 @@
-import PropTypes from 'prop-types';
 import React, {Component} from 'react';
-import {CheckBox, Icon, colors} from 'react-native-elements';
 import {View} from 'react-native';
+import {CheckBox, Icon, colors} from 'react-native-elements';
+
 import {applyTheme} from '../utility/utils';
-import {theme} from '../index';
+import {theme} from '../../App';
```
Comment (mahima-vs) [inline examples/src/components/Checkbox.js:6]: @saksham695 From where we are using theme now ?

File: examples/src/components/Checkbox.js [hunk 18]
```diff
@@ -1,64 +1,43 @@
...
+      checked: !this.state.checked,
+    });
+  };
+  render() {
     if (this.state.checked) {
-      this.setState({
-        checked: !this.state.checked,
-        color: uncheckedColor,
-      });
+      this.props.titleStyle.color = this.props.titleStyle.checkedColor
```
Comment (mahima-vs) [inline examples/src/components/Checkbox.js:33]: Conditional rendering is always not a good way to write a code. 
```
if(this.props.titleStyle.checkedColor){
   this.props.titleStyle.color = this.props.titleStyle.checkedColor
}
```
Use which one looks more better to you.

File: examples/src/components/Checkbox.js [hunk 19]
```diff
@@ -1,64 +1,43 @@
...
+        ? this.props.titleStyle.checkedColor
+        : this.props.titleStyle.color;
     } else {
-      this.setState({
-        checked: !this.state.checked,
-        color: checkedColor,
-      });
+      this.props.titleStyle.color = this.props.titleStyle.uncheckedColor
```
Comment (mahima-vs) [inline examples/src/components/Checkbox.js:37]: same here

File: examples/src/components/Checkbox.js [hunk 20]
```diff
@@ -1,64 +1,43 @@
-import PropTypes from 'prop-types';
 import React, {Component} from 'react';
-import {CheckBox, Icon, colors} from 'react-native-elements';
 import {View} from 'react-native';
```
Comment (mahima-vs) [inline examples/src/components/Checkbox.js:2]: @saksham695 Please sort lines if not done already.

File: examples/src/components/card.js [hunk 21]
```diff
@@ -1,54 +1,69 @@
...
-            {component.childrens.map((componentData, i) => {
-              return this.selectComponent(componentData, i);
-            })}
-          </View>
-        );
-      }
+  constructor(props) {
+    super(props);
+    if (theme) {
+      props = theme ? applyTheme(props, theme) : props;
```
Comment (mahima-vs) [inline examples/src/components/card.js:13]: Please improve @saksham695

File: examples/src/components/card.js [hunk 22]
```diff
@@ -1,54 +1,69 @@
...
-          return this.selectComponent(component, index);
-        })}
+        containerStyle={this.props.containerStyle}
+        dividerStyle={this.props.dividerStyle}
+        title={this.props.title}
+        titleStyle={this.props.titleStyle}
+        wrapperStyle={this.props.wrapperStyle}>
+        <ScrollView>
+          {this.props.childrens.size === 0 ? (
+            <Text>Add</Text>
```
Comment (mahima-vs) [inline examples/src/components/card.js:26]: ```            <Text>Add</Text> ```
What add ?

Comment (mahima-vs) [reply examples/src/components/card.js:26]: Does this works with ``` size ``` ?
@saksham695 ^^

File: examples/src/components/icon.js [hunk 23]
```diff
@@ -8,6 +8,12 @@ import {applyTheme} from '../utility/utils';
 import {theme} from '../index';
 
 export default class IconComponent extends Component {
+  constructor(props) {
+    super(props);
+    if (theme) {
+      this.props = theme ? applyTheme(this.props, theme) : this.props;
```
Comment (mahima-vs) [inline examples/src/components/icon.js:14]: @saksham695 Check the code.

File: examples/src/components/text.js [hunk 24]
```diff
@@ -9,7 +9,9 @@ import {theme} from '../index';
 export default class TextComponent extends React.Component {
   componentWillMount() {
     // Applying theme on text
-    this.props = theme ? applyTheme(this.props, theme) : this.props;
+    if (theme) {
+      this.props = theme ? applyTheme(this.props, theme) : this.props;
```
Comment (mahima-vs) [inline examples/src/components/text.js:13]: Check the code @saksham695

File: src/components/Checkbox.js [hunk 25]
```diff
@@ -3,7 +3,7 @@ import React, { Component } from "react";
 import { CheckBox, Icon, colors } from "react-native-elements";
 import { View } from "react-native";
 import { applyTheme } from "../utility/utils";
-import { theme } from "../../index";
+import { theme } from "../../App";
```
Comment (mahima-vs) [inline src/components/Checkbox.js:6]: ```import { theme } from "../../App";```
@saksham695 Pls fix .

File: examples/src/components/button.js [hunk 26]
```diff
@@ -9,13 +9,13 @@
...
+import {Button} from 'react-native-elements';
+import {View} from 'react-native';
 
-import { applyTheme } from '../utility/utils';
-import { shapeStyles } from '../style/buttonStyle';
-import { theme } from '../index';
+import {applyTheme} from '../utility/utils';
+import {shapeStyles} from '../style/buttonStyle';
+import {theme} from '../index';
+import {NavigationContext} from '@react-navigation/native';
```
Comment (mahima-vs) [inline examples/src/components/button.js:18]: Pls sort.

File: examples/src/components/card.js [hunk 27]
```diff
@@ -1,54 +1,77 @@
...
+        <ScrollView>
+          {this.props.childrens.length === 0 ? (
+            <Text
+              style={{
+                fontWeight: 'bold',
+                fontSize: 20,
+                marginTop: 100,
+                marginLeft: 30,
+              }}>
+              Please Add Your Component Inside Card
```
Comment (mahima-vs) [inline examples/src/components/card.js:33]: ```              Please Add Your Component Inside Card```
Move it to the constants file.

**Pull request merged**